drivers/pci/controller/pcie-brcmstb.c

```diff
@@ -1473,13 +1475,32 @@ static int brcm_pcie_start_link(struct brcm_pcie *pcie)
 	u16 nlw, cls, lnksta, tmp16;
 	bool ssc_good = false;
 	int ret, i;
+	u32 tmp;
 
 	/* Limit the generation if specified */
 	if (pcie->gen)
 		brcm_pcie_set_gen(pcie, pcie->gen);
 
 	/* Unassert the fundamental reset */
-	ret = pcie->cfg->perst_set(pcie, 0);
+	if (pcie->tperst_clk_ms) {
+		/*
+		 * Increase Tperst_clk time by forcing PERST# output low while
+		 * the internal reset is released, so the PLL generates stable
+		 * refclk output further in advance of PERST# deassertion.
+		 */
+		tmp = readl(pcie->base + HARD_DEBUG(pcie));
+		u32p_replace_bits(&tmp, 1, PCIE_MISC_HARD_PCIE_HARD_DEBUG_PERST_ASSERT_MASK);
+		writel(tmp, pcie->base + HARD_DEBUG(pcie));
+
+		ret = pcie->cfg->perst_set(pcie, 0);
+		msleep(pcie->tperst_clk_ms);
```

Comment:
What is the typical value for your sleep?
There were issues for short sleeps with msleep (<30ms), but I believe those are now resolved.

`fsleep(usecs)` is the recommended call - https://elixir.bootlin.com/linux/v6.13.3/source/Documentation/timers/delay_sleep_functions.rst (was Documentation/timers/timers-howto.rst)

Reply:
The worst-case EP delay (also the reason for implementing it) is 100ms and that is a minimum. I was unaware of fsleep.